File: batch-operations/javascript/node-postgres/src/occRetry.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

/**
 * OCC retry logic with exponential backoff for Aurora DSQL batch operations.
 */

class MaxRetriesExceededError extends Error {
  constructor(maxRetries) {
    super(`Max retries exceeded: failed after ${maxRetries} retries`);
    this.name = "MaxRetriesExceededError";
    this.maxRetries = maxRetries;
  }
}

function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

/**
 * Execute a database operation with OCC conflict retry and exponential backoff.
 *
 * Retries on SQLSTATE 40001 (serialization failure) with exponential backoff.
 * Each retry rolls back the current transaction before waiting.
 *
 * @param {import('pg').PoolClient} client - A node-postgres pool client.
 * @param {(client: import('pg').PoolClient) => Promise<*>} operation - Async
 *   function that performs database work. Should NOT commit.
 * @param {number} [maxRetries=3] - Maximum retry attempts.
 * @param {number} [baseDelay=0.1] - Base delay in seconds for backoff.
 * @returns {Promise<*>} The return value of `operation(client)`.
 */
async function executeWithRetry(client, operation, maxRetries = 3, baseDelay = 0.1) {
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await operation(client);
    } catch (e) {
      if (e.code === "40001") {
        await client.query("ROLLBACK");
        if (attempt >= maxRetries) {
          throw new MaxRetriesExceededError(maxRetries);
        }
        const delay = baseDelay * Math.pow(2, attempt);
        console.warn(
          `OCC conflict, retry ${attempt + 1}/${maxRetries} after ${delay.toFixed(1)}s`
        );
        await sleep(delay);
      } else {
        throw e;
      }
    }
  }
  throw new MaxRetriesExceededError(maxRetries);
}

module.exports = { MaxRetriesExceededError, executeWithRetry, sleep };

File: batch-operations/java/pgjdbc/src/main/java/com/example/dsql/OccRetry.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

package com.example.dsql;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.logging.Logger;

/**
 * OCC retry logic with exponential backoff for Aurora DSQL batch operations.
 */
public class OccRetry {

    private static final Logger logger = Logger.getLogger(OccRetry.class.getName());
    private static final String SERIALIZATION_FAILURE = "40001";

    @FunctionalInterface
    public interface DatabaseOperation<T> {
        T execute(Connection connection) throws SQLException;
    }

    public static class MaxRetriesExceededException extends Exception {
        private final int maxRetries;

        public MaxRetriesExceededException(int maxRetries) {
            super("Max retries exceeded: failed after " + maxRetries + " retries");
            this.maxRetries = maxRetries;
        }

        public int getMaxRetries() {
            return maxRetries;
        }
    }

    /**
     * Execute a database operation with OCC conflict retry and exponential backoff.
     *
     * @param connection a JDBC connection (autoCommit should be false)
     * @param operation  the database operation to execute
     * @param maxRetries maximum retry attempts (default 3)
     * @param baseDelay  base delay in seconds for backoff (default 0.1)
     * @return the return value of the operation
     */
    public static <T> T executeWithRetry(
            Connection connection,
            DatabaseOperation<T> operation,
            int maxRetries,
            double baseDelay) throws MaxRetriesExceededException, SQLException {

        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                return operation.execute(connection);
            } catch (SQLException e) {
                if (SERIALIZATION_FAILURE.equals(e.getSQLState())) {
                    connection.rollback();
                    if (attempt >= maxRetries) {
                        throw new MaxRetriesExceededException(maxRetries);
                    }
                    double delay = baseDelay * Math.pow(2, attempt);
                    logger.warning(String.format(
                            "OCC conflict, retry %d/%d after %.1fs",
                            attempt + 1, maxRetries, delay));
                    try {
                        Thread.sleep((long) (delay * 1000));
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new SQLException("Retry interrupted", ie);
                    }
                } else {
                    throw e;
                }
            }
        }
        throw new MaxRetriesExceededException(maxRetries);
    }

    public static <T> T executeWithRetry(
            Connection connection,
            DatabaseOperation<T> operation) throws MaxRetriesExceededException, SQLException {
        return executeWithRetry(connection, operation, 3, 0.1);
    }
}
